src/main/java/org/osc/controller/nsc/api/SampleSdnRedirectionApi.java

```diff
@@ -0,0 +1,334 @@
+
+        InspectionHookEntity retValEntity = this.txControl.required(() -> {
+            InspectionPortEntity inspectionPortTmp =(InspectionPortEntity) getInspectionPort(inspectionPort);
+            throwExceptionIfNullEntity(inspectionPortTmp, inspectionPort);
+
+            InspectionHookEntity inspectionHookEntity = this.utils.findInspHookByInspectedAndPort(inspected, inspectionPortTmp);
+
+            if (inspectionHookEntity == null) {
+                inspectionHookEntity = this.utils.makeInspectionHookEntity(inspected, inspectionPortTmp, tag,
+                                                                           encType, order, failurePolicyType);
+            }
+
+            return this.em.merge(inspectionHookEntity);
+        });
+
+        return retValEntity.getHookId();
+    }
+
+    private void throwExceptionIfNullEntity(InspectionPortEntity inspectionPortTmp, InspectionPortElement inspectionPort)
+        throws IllegalArgumentException {
+        if (inspectionPortTmp == null) {
+            String ingressId = inspectionPort.getIngressPort() != null ? inspectionPort.getIngressPort().getElementId() : null;
+            String egressId = inspectionPort.getEgressPort() != null ? inspectionPort.getEgressPort().getElementId() : null;
+            String msg = String.format("Cannot find inspection port for inspection hook "
+                    + "id: %s; ingressId: %s; egressId: %s\n", inspectionPort.getElementId(), ingressId, egressId);
+            throw new IllegalArgumentException(msg);
+        }
+    }
+
+    @Override
+    public void removeInspectionHook(List<NetworkElement> inspectedPort, InspectionPortElement inspectionPort)
+            throws Exception {
+
+        if (inspectedPort != null && inspectedPort.size() > 0) {
+
+            NetworkElement inspected = inspectedPort.get(0);
+
+            this.txControl.required(() -> {
+                InspectionHookEntity inspectionHook = this.utils.findInspHookByInspectedAndPort(inspected, inspectionPort);
+                if (inspectionHook != null) {
+                    this.utils.removeSingleInspectionHook(inspectionHook);
+                }
+                return null;
+            });
+        }
+    }
+
+    @Override
+    public Long getInspectionHookTag(NetworkElement inspectedPort, InspectionPortElement inspectionPort)
+            throws NetworkPortNotFoundException, Exception {
+        InspectionHookElement inspectionHook = getInspectionHook(inspectedPort, inspectionPort);
+        return inspectionHook == null ? null : inspectionHook.getTag();
+    }
+
+    @Override
+    public void setInspectionHookTag(NetworkElement inspectedPort, InspectionPortElement inspectionPort, Long tag)
+            throws Exception {
+
+        this.txControl.required(() -> {
+            InspectionHookEntity inspectionHook = this.utils.findInspHookByInspectedAndPort(inspectedPort, inspectionPort);
+            if (inspectionHook != null) {
```

Comment:
looks like lots of places where we silently fail. Looking at the previous code, we used to fail with an NPE, thats no longer the case here. No log messages either, so will be hard to debug. Please review to be consistent with the previous code or add logging etc

```
InspectionHook inspectionHook = neutronApi.getInspectionHookByPorts(this.region, inspectedPort.getElementId(), inspectionPort);
-            inspectionHook.setTag(tag);
```